examples/abitest/module_linear_handles/rust/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+/// Tests the linear-handles feature in `oak_io`.
+// Note: This environment is very bare. the Oak SDK does not support linear handles, so we can
+// only use the ABI and `oak_io`
+use oak_abi::proto::oak::label::Label;
+use oak_io::{
+    handle::{ReadHandle, WriteHandle},
+    OakStatus,
+};
+
+fn run_tests() {
+    test_cloned_handles_are_valid_and_distinct();
+
+    test_original_and_clone_are_connected();
+
+    test_drop_closes_handle();
+}
+
+fn test_cloned_handles_are_valid_and_distinct() {
+    let (w, r) = channel_create().expect("Failed to create channel");
+    let w2 = w.clone();
+    let r2 = r.clone();
+
+    // Check that the handles are not invalid
+    assert!(w2.handle != oak_abi::INVALID_HANDLE);
+    assert!(r2.handle != oak_abi::INVALID_HANDLE);
+    // Check that the handles are distinct from the original ones
+    assert!(w.handle != w2.handle);
+    assert!(r.handle != r2.handle);
+}
+
+// Writes from the cloned handle should be readable from the original, and vice versa
+fn test_original_and_clone_are_connected() {
+    let (w, r) = channel_create().expect("Failed to create channel");
+    let w2 = w.clone();
+    let r2 = r.clone();
+
+    // Write from original, read from clone
+    channel_write(&w, "msg1").expect("Failed to write to w");
+    let msg1 = channel_read(&r2).expect("Failed to read from r2");
+    assert_eq!(msg1, "msg1");
+
+    // Write from clone, read from original
+    channel_write(&w2, "msg2").expect("Failed to write to w2");
+    let msg2 = channel_read(&r).expect("Failed to read from r");
+    assert_eq!(msg2, "msg2");
+}
+
+fn test_drop_closes_handle() {
+    let (w, r) = channel_create().expect("Failed to create channel");
+    assert_eq!(channel_read(&r), Err(OakStatus::ErrChannelEmpty));
+
+    // Explicitly drop the handle
+    core::mem::drop(w);
+
+    // Channel is now orphaned
+    assert_eq!(channel_read(&r), Err(OakStatus::ErrChannelClosed));
+}
+
+#[no_mangle]
+pub extern "C" fn linear_handles_oak_main(init_handle: u64) {
+    let _ = std::panic::catch_unwind(|| {
+        // Set up a panic handler that reports errors to the parent node
+        std::panic::set_hook(Box::new(move |panic_info| {
```

Comment:
Just checking: Are you defining this hook instead of using [`set_panic_hook` in the SDK](https://github.com/project-oak/oak/blob/a5dcfe403e2a0b4d6273c70d6973b57955f14dff/sdk/rust/oak/src/lib.rs#L476) because the `SDK` does not support linear-handles, and so cannot be used in this test?

Reply:
Yes, that's exactly right 😄